Add `track_format` config option

Bettehem opened the pull request:
Hello again!
I added a new config option: `tracklist_formatting`.
This solves #730 and #558

hrkfdn commented:
Thanks for the PR, appreciate it. I think a more flexible/customizable approach would be the one described here: https://github.com/hrkfdn/ncspot/issues/558

It would allow custom formats for tracks (and maybe albums and other types) so users can decide for themselves, but is probably a little bit more complex.

hrkfdn commented:
Thanks for the PR, appreciate it. I think a more flexible/customizable approach would be the one described here: https://github.com/hrkfdn/ncspot/issues/558

It would allow custom formats for tracks (and maybe albums and other types) so users can decide for themselves, but is probably a little bit more complex.

Bettehem commented:
Oh yeah, that sounds like a good idea! I could try changing it to work like that🤔

Bettehem commented:
Hello, I have made some some modifications now. I added an `active_fields` config option which lets you configure what should be shown and in which column. This also removes the need for a separate `album_column` option. I also added some examples to the README. This should now solve both #730 and #558

hrkfdn commented on src/model/track.rs:
```diff
@@ -159,19 +161,105 @@ impl From<&SavedTrack> for Track {
-        write!(f, "{} - {}", self.artists.join(", "), self.title)
+        let active_fields = config::ACTIVE_FIELDS.read().unwrap().clone();
+        let mut fields: Vec<ActiveField> = Vec::new();
+        fields.append(
+            &mut active_fields
+                .left
+                .unwrap_or(vec![ActiveField::Artists, ActiveField::Title])
+                .clone(),
+        );
+        fields.append(
+            &mut active_fields
+                .center
+                .unwrap_or(vec![ActiveField::Album])
+                .clone(),
+        );
+        if fields.contains(&ActiveField::Title) && !fields.contains(&ActiveField::Artists) {
+            write!(f, "{}", self.title)
+        } else if fields.contains(&ActiveField::Artists) && !fields.contains(&ActiveField::Title) {
+            write!(f, "{}", self.artists.join(", "))
+        } else if !fields.contains(&ActiveField::Default)
+            && !fields.contains(&ActiveField::Title)
+            && !fields.contains(&ActiveField::Artists)
+        {
+            write!(f, "")
+        } else {
+            let mut title_index = 1;
+            let mut artists_index = 0;
+            for (i, field) in fields.iter().enumerate() {
+                match field {
+                    config::ActiveField::Title => title_index = i,
+                    config::ActiveField::Artists => artists_index = i,
+                    _ => {}
+                }
+            }
+            if title_index < artists_index {
+                write!(f, "{} - {}", self.title, self.artists.join(", "))
+            } else {
+                write!(f, "{} - {}", self.artists.join(", "), self.title)
+            }
```
There is a lot of logic here to generate the output string for the columns. I'm wondering if there is a more dynamic approach, i.e. the user can provide something like `format_left = "%artists - %title"` which is then replaced.

Bettehem replied:
Ok, I like the idea. It would probably make the logic simpler. I'll see what I can do!

hrkfdn reviewed (commented):
Thanks for reiterating and tackling this. I have a suggestion in the comment below, it would also apply to the other columns. What do you think?

Bettehem commented:
Hi, Sorry for taking so long, I have been busy with other stuff, but I have now made a more flexible system for customizing the formatting. What do you think?

hrkfdn commented:
Thanks, looks pretty good already. I've rebased the latest `main` and did a small refactoring. I'm wondering if we should support formatting for podcast episodes as well? I think they may stick out if you have a playlist with a mix of tracks and podcasts..

Bettehem commented:
Oh, true. I didn't know you could mix podcasts and songs in the same playlist 🤦
But yeah the formatting should probably work for podcast episodes too if they can be added to the same list.
Thanks for the feedback again, I'll try to make it work for episodes too!

Bettehem commented:
Hmm, I was just testing, and noticed that the custom formatting already works for Episodes in playlists, because episodes in playlists are being treated as tracks. I made a new commit which moves the custom format function to Playable impl so if in the future playlists differentiate between Episodes and Playlists its easier to implement

hrkfdn commented:
> Hmm, I was just testing, and noticed that the custom formatting already works for Episodes in playlists, because episodes in playlists are being treated as tracks.

Huh, indeed. Sounds like a bug, I've just checked and it looks like either rspotify or the Spotify API returns them all as tracks. I have reported this upstream at rspotify: https://github.com/ramsayleung/rspotify/issues/320

> I made a new commit which moves the custom format function to Playable impl so if in the future playlists differentiate between Episodes and Playlists its easier to implement.

That makes sense, thanks.

I think we can merge this now. I was wondering whether we should rename the config variables to `track_formatting` as it's a little bit shorter and reflects the affected playable type better. What do you think?

Once the bug is fixed we can also support custom podcast formatting, which may have different fields. Also in the future we could add support custom formatting for other playable types, i.e. albums.

hrkfdn commented:
I have changed the naming and merged it. Thanks for your contribution and considering all the annoying comments :smile:

Bettehem commented:
> I have changed the naming and merged it. Thanks for your contribution and considering all the annoying comments 😄

Nice, and thank you for taking the time to give feedback and suggestions, I'm still a Rust noob so it has been very helpful!
